docs: unify module pages (#725)
denoland/docs: runtime/fundamentals/modules.md

---
title: "Using modules in Deno"
oldUrl:
  - /runtime/manual/basics/modules/
  - /runtime/manual/basics/modules/integrity_checking/
  - /runtime/manual/basics/modules/module_metadata/
  - /runtime/manual/basics/modules/publishing_modules/
  - /runtime/manual/basics/modules/reloading_modules/
  - /runtime/manual/basics/vendoring/
  - /runtime/manual/advanced/http_imports/
  - /runtime/manual/advanced/publishing/dnt/
  - /runtime/manual/advanced/publishing/
  - /runtime/manual/examples/manage_dependencies
  - /runtime/manual/node/cdns.md
  - /runtime/manual/linking_to_external_code
  - /runtime/manual/linking_to_external_code/reloading_modules
  - /runtime/fundamentals/esm.sh
---

Deno uses
[ECMAScript modules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules)
as its default module system to align with modern JavaScript standards and to
promote a more efficient and consistent development experience. It's the
official standard for JavaScript modules, allows for better tree-shaking,
improved tooling integration, and native support across different environments.

> By adopting ECMAScript modules, Deno ensures compatibility with the evolving

Author: This sentence is loooooong and a bit unwieldy, I'd suggest splitting it up a bit.

By adopting ECMAScript modules, Deno ensures compatibility with the ever-evolving JavaScript ecosystem. For developers, this means a streamlined and predictable module system that avoids the complexities associated with legacy module formats like CommonJS.

Reviewer: Good point, that reads much more natural. Looks like the German came through in me!

JavaScript ecosystem, providing developers with a streamlined and predictable
module system that eliminates the complexities associated with legacy module
formats like CommonJS.

## Importing modules

In this example the `add` function is imported from a local `calc.ts` module.

```ts title="calc.ts"
export function add(a: number, b: number): number {
  return a + b;
}
```

```ts title="main.ts"
// imports the `calc.ts` module next to this file
import { add } from "./calc.ts";

console.log(add(1, 2)); // 3
```

You can run this example by calling `deno run main.ts` in the directory that
contains both `main.ts` and `calc.ts`.

> With ECMAScript modules, local import specifiers must always included the full

Author: ```suggestion
With ECMAScript modules, local import specifiers must always include the full
```

file extension. It cannot be omitted.

```ts title="example.ts"
// WRONG: missing file extension
import { add } from "./calc";

// CORRECT: includes file extension
import { add } from "./calc.ts";
```

## Importing third party modules and libraries

> Using third modules in Deno uses the same `import` syntax as your local code.

Author: ```suggestion
When working with third-party modules in Deno, use the same `import` syntax as you do for local code.
```

Third party modules are typically imported from a remote registry and start with
`jsr:` , `npm:` or `https://` for URL imports.

```ts title="main.ts"
import { camelCase } from "jsr:@luca/cases@1.0.0";
import { say } from "npm:cowsay@1.6.0";
import { pascalCase } from "https://deno.land/x/case/mod.ts";
```

> You'll find most Deno-related code on [JSR](https://jsr.io), which is the
> recommended registry to use with Deno.

Author: ```suggestion
Deno recommends [JSR](https://jsr.io), the modern JavaScript registry, for third party modules. There, you'll find plenty of well documented ES modules for your projects, including the [Deno Standard Library](/runtime/fundamentals/standard_library/).
```

## Managing third party modules and libraries

> Typing out the module with the full version specifier can become tedious real
> fast when you're importing them from multiple files. We can centralize the
> management of remote modules via the `imports` key in `deno.json`. The `imports`
> key is used to remap specifiers. We call this the import map.
> 

Author: ```suggestion
Typing out the module name with the full version specifier can become tedious when importing them in multiple files. You can centralize management of remote modules with an `imports` field in your `deno.json` file. We call this `imports` field the **import map**.

```

```json title="deno.json"
{
  "imports": {
    "@luca/cases": "jsr:@luca/cases@^1.0.0",
    "cowsay": "npm:cowsay@^1.6.0",
    "cases": "https://deno.land/x/case/mod.ts"
  }
}
```

> With the remapped specifiers our code gets a bit simpler:

Author: ```suggestion
With remapped specifiers, the code looks cleaner:
```

```ts title="main.ts"
import { camelCase } from "@luca/cases";
import { say } from "cowsay";
import { pascalCase } from "cases";
```

The remapped name can be any valid specifier. It's a very powerful feature in
Deno that can remap anything. Learn more about everything the import map can do
[here](/runtime/manual/basics/import_maps/).

## Adding dependencies with deno add

> The installation process is made easier with the `deno add` subcommand. It will
> automatically add the latest version of the package you requested to the
> `imports` section in `deno.json`.

Author: ```suggestion
The installation process is made easy with the `deno add` subcommand. It will
automatically add the latest version of the package you requested to the
`imports` section in `deno.json`.
```

| `1.2.x`   | Any patch version within the minor version 1.2. For example, `1.2.0`, `1.2.1`, etc.                                                                                 | `1.2.x`   |
| `1.x`     | Any minor and patch version within the major version 1. For example, `1.0.0`, `1.1.0`, `1.2.0`, etc.                                                                | `1.x`     |
| `*`       | Any version is allowed.                                                                                                                                             | `*`       |

## URL imports

Deno also supports import statements that reference URLs, either directly:

```js
import { Application } from "https://deno.land/x/oak/mod.ts";
```

or part of your `deno.json` import map:

```json
{
  "imports": {
    "oak": "https://deno.land/x/oak/mod.ts"
  }
}
```

Supporting URL imports enables us to support the following JavaScript CDNs, as
they provide URL access to JavaScript modules:

- [deno.land/x](https://deno.land/x)
- [esm.sh](https://esm.sh)
- [unpkg.com](https://unpkg.com)

> URL imports are useful if you have a small, often single file, Deno project that
> doesn't require any other configuration because you can avoid having a
> `Deno.json` file at all. It's not advised to use this style of import in larger
> applications as you may end up with version conflicts (where different files use
> different version specifiers).

Author: ```suggestion
URL imports are useful if you have a small, often single file, Deno project that
doesn't require any other configuration. With URL imports, you can avoid having a
`Deno.json` file at all. It is **not** advised to use this style of import in larger
applications however, as you may end up with version conflicts (where different files use different version specifiers).
```

:::info

Use URL imports with caution, and only **from trusted sources**. If the server
is compromised, it could serve malicious code to your application. They can also
cause versioning issues if you import different versions in different files. URL
imports remain supported, **but we recommend using a package registry for the
best experience.**

:::

### Overriding URL imports

The other situation where import maps can be very useful is to override URL
imports in specific modules.

> Let's say you want to override a `https://deno.land/x/my-library@1.0.0/mod.ts`

Author: can we use a jsr example rather than deno.land?

Reviewer: URL imports cannot be overridden with the `scopes` import feature afaik

Deno uses a `deno.lock` file to check external module integrity. To opt into a
lock file, either:

1. Create a `deno.json` file in the current or an ancestor directory, which will
   automatically create an additive lockfile at `deno.lock`.
2. Use the `--lock=deno.lock` flag to enable and specify lock file checking. To
   update or create a lock use `--lock=deno.lock --frozen=false`. The
   `--lock=deno.lock` tells Deno what the lock file to use is, while the
   `--frozen=false` is used to output dependency hashes to the lock file.

A `deno.lock` might look like this, storing a hash of the file against the
dependency:

```json title="deno.lock"
{
  "https://deno.land/std@0.224.0/textproto/mod.ts": "3118d7a42c03c242c5a49c2ad91c8396110e14acca1324e7aaefd31a999b71a4",
  "https://deno.land/std@0.224.0/io/util.ts": "ae133d310a0fdcf298cea7bc09a599c49acb616d34e148e263bcb02976f80dee",
  "https://deno.land/std@0.224.0/async/delay.ts": "35957d585a6e3dd87706858fb1d6b551cb278271b03f52c5a2cb70e65e00c26a",
   ...
}
```

### Auto-generated lockfile

As mentioned above, when a Deno configuration file is resolved (ex. `deno.json`)

> then an additive lockfile will be automatically generated. By default, the path

Author: ```suggestion
then a lockfile will be automatically generated. By default, the path
```

> As mentioned above, when a Deno configuration file is resolved (ex. `deno.json`)
> then an additive lockfile will be automatically generated. By default, the path
> of this lockfile will be `deno.lock`. You can change this path by updating your
> `deno.json` to specify this:
> 
> ```json title="deno.json"
> {
>   "lock": "./lock.file"
> }
> ```

Author: ```suggestion
As mentioned above, when a Deno configuration file is resolved (eg. `deno.json`), a lockfile will be automatically generated. By default, the path
of this lockfile will be in the directory root - `deno.lock`. You can change this path by updating your
`deno.json` to specify this:

```json title="deno.json"
{
  "lock": "./lock.file"
}
```
```

```json title="deno.json"
{
  "lock": false
}
```

### Using `--lock` and `--frozen=false` flags

You may have a file that imports a dependency and looks something like this:

```ts title="src/deps.ts"
export { xyz } from "https://unpkg.com/xyz-lib@v0.9.0/lib.ts";
```

To create a lock file, you can use the `--lock` and `--frozen=false` flags:

```shell
# Create/update the lock file "deno.lock".
deno cache --lock=deno.lock --frozen=false src/deps.ts

# Include it when committing to source control.
git add -u deno.lock
git commit -m "feat: Add support for xyz using xyz-lib"
git push
```

Collaborator on another machine -- in a freshly cloned project tree:

```shell
# Download the project's dependencies into the machine's cache, integrity
# checking each resource.
deno cache --reload --lock=deno.lock src/deps.ts

# Done! You can proceed safely.
deno test --allow-read src
```

### Runtime verification

Like caching above, you can also use lock files during use of the `deno run` sub
command, validating the integrity of any locked modules during the run. Remember
that this only validates against dependencies previously added to the lock file.

> You can take this a step further as well by using the `--cached-only` flag to

Author: ```suggestion
You can take this a step further by using the `--cached-only` flag to
```

---
title: "Using modules in Deno"
oldUrl: 
  - /runtime/manual/basics/modules/
  - /runtime/manual/basics/modules/integrity_checking/
  - /runtime/manual/basics/modules/module_metadata/
  - /runtime/manual/basics/modules/proxies/
  - /runtime/manual/basics/modules/publishing_modules/
  - /runtime/manual/basics/modules/reloading_modules/
  - /runtime/manual/advanced/publishing/dnt/
  - /runtime/manual/advanced/publishing/
  - /runtime/manual/basics/vendoring/
  - /runtime/manual/node/cdns.md
  - /runtime/manual/advanced/http_imports/
---

Deno standardizes the way modules are imported in both JavaScript and TypeScript
using the
[ECMAScript module standard](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules)

> by default. In line with this standard, **file names must be specified in

Reviewer 2: "Deno standardizes the way modules are imported..." is odd phrasing

How about "Deno uses ECMAScript modules exclusively. That is, `import` rather than `require`."

Maybe add a bit of context about why -- something like "since 2015(?) ES Modules have been part of the specification that defines JavaScript. As such ES Modules work in the browser, CommonJS does not. Deno's goal is to narrow the gap between browser and server."

full**. You should not omit the file extension and there is no special handling

> of `index.js`.

Reviewer 2: This belongs in its own section - not part of the introduction. Also this is specific to local modules - it does not apply to https modules, npm modules, or jsr modules.

Something like:

> ## Local imports
> 
> To link to code in your project, use relative imports. These always start with "./" or "../" and end in a file extension. For example... 

After this section about local imports then add a section about remote imports aka third party modules - third party modules always start like this: `https://` or `npm:` or `jsr:`

Once these are explained, then talk about bare specifiers and import maps. Followed by more detailed information.

import { render } from "https://esm.sh/preact";
```

You can also import modules from a URL by adding it to your `deno.json` import
map:

```json title="deno.json"
{
  "imports": {
    "preact": "https://esm.sh/preact"
  }
}
```

URL imports should be used with caution, as they can introduce security risks.
When importing modules from a URL, you are trusting the server to serve the
correct code. If the server is compromised, it could serve malicious code to
your application. For this reason, it is recommended to **use URL imports only
from trusted sources**. They can also cause versioning issues if you import
different versions in different files.

## Package Versions

It is possible to specify a version range for the package you are importing.
This is done using the `@` symbol followed by a version range specifier, and
follows the [semver](https://semver.org/) versioning scheme.

```bash
deno add @scopename/mypackage           # latest version
deno add @scopename/mypackage@16.1.0    # exact version
deno add @scopename/mypackage@^16.1.0   # latest patch version 16.x
deno add @scopename/mypackage@~16.1.0   # latest version that doesn't increment the first non-zero portion

> ```

Reviewer 2: It's not just with `deno add`, also within specifiers.

The manual should explain the syntax of version ranges https://gist.github.com/jonlabelle/706b28d50ba75bf81d40782aa3c84b3e

imports in specific modules.

Let's say you want to override a `https://deno.land/x/my-library@1.0.0/mod.ts`
specifier that is used inside files coming from `https://deno.land/x/example/`
to a local patched version. You can do this by using a scope in the import map
that looks something like this:

```json
{
  "imports": {
    "example/": "https://deno.land/x/example/"
  },
  "scopes": {
    "https://deno.land/x/example/": {
      "https://deno.land/x/my-library@1.0.0/mod.ts": "./patched/mod.ts"
    }
  }
}
```

:::note

URL imports have no notion of packages. Only the import map at the root of your
project is used. Import maps used inside URL dependencies are ignored.

:::

## Proxies

Deno is able to handle network requests through a proxy server, useful for
various reasons such as security, caching, or accessing resources behind a
firewall. The runtime supports supports proxies for module downloads and the Web
standard `fetch` API.

Deno reads proxy configuration from environment variables: `HTTP_PROXY`,
`HTTPS_PROXY` and `NO_PROXY`.

In case of Windows, if environment variables are not found, Deno falls back to

> reading proxies from the registry.

Reviewer 2: This seems out of place. The proxy configuration works for modules but also for `fetch`. 

Maybe it makes sense here if its also written in the env var reference docs.

## Publishing modules

Any Deno program that defines an export can be published as a module. This
allows other developers to import and use your code in their own projects.
Modules can be published to [JSR](https://jsr.io), the modern JavaScript and
TypeScript registry. Check out the
[JSR documentation on publishing modules](https://jsr.io/docs/publishing-packages)
for more information.

## Publishing for Node.js

You can make your Deno modules available to Node.js users with the
[dnt](https://github.com/denoland/dnt) build tool.

dnt allows you to develop your Deno module mostly as-is and use a single Deno
script to build, type check, and test an npm package in an output directory.
Once built, you only need to `npm publish` the output directory to distribute it
to Node.js users.

For more details, see

> [https://github.com/denoland/dnt](https://github.com/denoland/dnt).

Reviewer 2: Modules can be published either to JSR or NPM or accessed via HTTPS (deno.land/x being one option). This needs to be reworked.

## Reloading modules

By default, Deno uses a global cache directory (`DENO_DIR`) for downloaded
dependencies. This cache is shared across all projects.

You can force deno to refetch and recompile modules into the cache using the
`--reload` flag of the `deno cache` or `deno run` subcommand.

```bash
# Reload everything
deno cache --reload my_module.

# Reload a specific module
deno cache --reload=jsr:@std/fs my_module.ts

> ```

Reviewer 2: Maybe call this section "Flushing cached dependencies" 

the examples are better if they use `deno run`
